File: src/net/eduard/api/lib/menu/MenuButton.java
package net.eduard.api.lib.menu;

import org.bukkit.inventory.ItemStack;

import net.eduard.api.lib.game.ClickEffect;

public class MenuButton  extends Slot{

	private String name="Botao";
	private int page = 1;
	private Menu menu;
	private transient Menu parentMenu;





	public MenuButton() {

	}
	public void setMenu(Menu menu) {
		this.menu = menu;
	}
	public MenuButton(Menu parent) {
		setParentMenu(parent);
		getParentMenu().addButton(this);

	}
	public MenuButton(String name) {
		setName(name);

	}
	public MenuButton(ItemStack icon) {

		setIcon(icon);

	}

	public MenuButton(String name, ItemStack icon) {
		setName(name);
		setIcon(icon);

	}

	public Menu getMenu() {
		return menu;
	}
	public Shop getShop() {
		return (Shop) menu;
	}
	private transient ClickEffect click;

	public ClickEffect getClick() {
		return click;
	}

	public void setClick(ClickEffect click) {
		this.click = click;
	}

	/**
	 * Retorna getItem()
	 * @return
	 */
	public ItemStack getIcon() {
		return getItem();
	}
	/**
	 * Executa o setItem()
	 * @param icon
	 */
	public void setIcon(ItemStack icon) {
		setItem(icon);
	}

	public boolean isCategory() {
		return menu!=null;
	}

	

	public int getPage() {
		return page;
	}

	public void setPage(int page) {
		this.page = page;
	}



	public String getName() {
		
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}


	public Menu getParentMenu() {
		return parentMenu;
	}

	public void setParentMenu(Menu parentMenu) {
		this.parentMenu = parentMenu;
	}
}

File: src/net/eduard/api/lib/menu/MenuButton.kt
package net.eduard.api.lib.menu

import net.eduard.api.lib.game.ClickEffect
import net.eduard.api.lib.kotlin.player
import net.eduard.api.lib.modules.Copyable
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

open class MenuButton(var name: String = "Botao"
                      , @Transient
                      var parentMenu: Menu? = null
                      , positionX: Int = 0,
                      positionY: Int = 1,
                      var page: Int = 1,
                      block: (MenuButton.() -> Unit)? = null

) : Slot(positionX, positionY) {



    var menu: Menu? = null

    init {
        block?.invoke(this)
        parentMenu?.addButton(this)

    }

    constructor(parent: Menu) : this(parentMenu = parent)


    val shop: Shop
        get() = menu as Shop

    @Transient
    var click: ClickEffect? = null

    open fun getIcon(player: Player): ItemStack {
        return this.icon
    }

    var icon: ItemStack
        get() = item!!
        set(icon) {
            item = icon
        }

    val isCategory: Boolean
        get() = menu != null

}